File: apps/api/src/runsight_api/transport/schemas/workflows.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, StrictBool


class CanvasViewport(BaseModel):
    x: float = 0.0
    y: float = 0.0
    zoom: float = 1.0


class WorkflowCanvasState(BaseModel):
    nodes: List[Dict[str, Any]] = Field(default_factory=list)
    edges: List[Dict[str, Any]] = Field(default_factory=list)
    viewport: CanvasViewport = Field(default_factory=CanvasViewport)
    selected_node_id: Optional[str] = None
    canvas_mode: Literal["dag", "state-machine"] = "dag"


class WorkflowHealthMetrics(BaseModel):
    run_count: int = 0
    eval_pass_pct: float | None = None
    eval_health: Literal["success", "warning", "danger"] | None = None
    total_cost_usd: float = 0.0
    regression_count: int = 0


class WarningItem(BaseModel):
    message: str
    source: str | None = None
    context: str | None = None

    model_config = {"json_schema_extra": {"additionalProperties": False}}


class WorkflowResponse(BaseModel):
    kind: Literal["workflow"]
    id: str
    name: Optional[str] = None
    description: Optional[str] = None
    yaml: Optional[str] = None
    canvas_state: Optional[WorkflowCanvasState] = None
    valid: bool = True
    validation_error: Optional[str] = None
    warnings: List[WarningItem] = Field(
        default_factory=list,
        json_schema_extra={"default": []},
    )
    block_count: int = 0
    modified_at: float | None = None
    enabled: bool = False
    commit_sha: str | None = None
    health: WorkflowHealthMetrics = Field(default_factory=WorkflowHealthMetrics)


class WorkflowListResponse(BaseModel):
    items: List[WorkflowResponse]
    total: int


class WorkflowCreate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    yaml: str
    canvas_state: Optional[WorkflowCanvasState] = None
    commit: bool = True


class WorkflowUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    yaml: str
    canvas_state: Optional[WorkflowCanvasState] = None


class WorkflowCommitCreate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    yaml: str
    canvas_state: Optional[WorkflowCanvasState] = None
    message: str = Field(min_length=1)


class WorkflowCommitResponse(BaseModel):
    hash: str
    message: str


class WorkflowSimulationCreate(BaseModel):
    yaml: str


class WorkflowSimulationResponse(BaseModel):
    branch: str
    commit_sha: str


class WorkflowDeleteResponse(BaseModel):
    id: str
    deleted: bool
    runs_deleted: int


class WorkflowEnabledUpdate(BaseModel):
    enabled: StrictBool

File: apps/gui/src/types/schemas/workflows.ts
import { z } from "zod";

const CanvasStateSchema = z.object({
  nodes: z.array(z.record(z.any())).default([]),
  edges: z.array(z.record(z.any())).default([]),
  viewport: z.object({
    x: z.number().default(0),
    y: z.number().default(0),
    zoom: z.number().default(1),
  }).default({ x: 0, y: 0, zoom: 1 }),
  selected_node_id: z.string().nullable().default(null),
  canvas_mode: z.enum(["dag", "state-machine"]).default("dag"),
});

export const WorkflowResponseSchema = z.object({
  id: z.string(),
  name: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  blocks: z.record(z.any()).default({}),
  edges: z.array(z.record(z.any())).default([]),
  canvas_state: CanvasStateSchema.optional(),
  status: z.string().optional(),
  updated_at: z.string().optional(),
  created_at: z.string().optional(),
  last_run_duration: z.number().optional(),
  last_run_cost_usd: z.number().optional(),
  last_run_completed_at: z.string().optional(),
  step_count: z.number().optional(),
  block_count: z.number().optional(),
});
export type WorkflowResponse = z.infer<typeof WorkflowResponseSchema>;

export const WorkflowListResponseSchema = z.object({
  items: z.array(WorkflowResponseSchema),
  total: z.number(),
});
export type WorkflowListResponse = z.infer<typeof WorkflowListResponseSchema>;

export const WorkflowCreateSchema = z.object({
  id: z.string().optional(),
  name: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  blocks: z.record(z.any()).default({}),
  edges: z.array(z.record(z.any())).default([]),
  canvas_state: CanvasStateSchema.optional(),
});
export type WorkflowCreate = z.infer<typeof WorkflowCreateSchema>;

export const WorkflowUpdateSchema = z.object({
  name: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  blocks: z.record(z.any()).nullable().optional(),
  edges: z.array(z.record(z.any())).nullable().optional(),
  canvas_state: CanvasStateSchema.nullable().optional(),
});
export type WorkflowUpdate = z.infer<typeof WorkflowUpdateSchema>;
